src/shogun/io/ARFFFile.h

```diff
@@ -0,0 +1,569 @@
+		 *
+		 */
+		void read();
+
+		/**
+		 * Returns string parsed in @relation line
+		 * @return the relation string
+		 */
+		std::string get_relation() const noexcept
+		{
+			return m_relation;
+		}
+
+		/**
+		 * Returns the name of the features parsed in @attribute
+		 * @return the relation string
+		 */
+		std::vector<std::string> get_feature_names() const noexcept
+		{
+			return m_attribute_names;
+		}
+
+		/**
+		 * Get combined features from parsed data
+		 * @return
+		 */
+		std::shared_ptr<CCombinedFeatures> get_features() const;
+
+	private:
+		/**
+		 * Processes a chunk. A chunk is defined as a set of lines that
+		 * are processed in the same way. A chunk ends when the func
+		 * sets the internal m_state to false.
+		 * Parsing can also end when the stream reaches EOF.
+		 *
+		 * @tparam LambdaT type of processing function
+		 * @tparam CheckT type of check function
+		 * @param func processing function that reads each line
+		 * @param check_func function that checks the result from the processing
+		 * function
+		 * @param skip_first whether to stream the first line
+		 */
+		template <typename LambdaT, typename CheckT>
+		void process_chunk(LambdaT&& func, CheckT&& check_func, bool skip_first)
+		{
+			m_state = false;
+
+			if (skip_first && !m_stream->eof())
+				func();
+
+			while (!m_state && !m_file_done)
+			{
+				consume_line(func);
+			}
+			if (!check_func())
+			{
+				SG_SERROR(
+				    "Parsing error on line %d: %s\n", m_line_number,
+				    m_current_line.c_str());
+			}
+		}
+
+		/**
+		 * Function called by process_chunk to process a "chunk" line by line.
+		 * This function also checks if EOF has been reached.
+		 *
+		 * @tparam T type of processing function
+		 * @param func line processing function
+		 */
+		template <typename T>
+		void consume_line(T&& func)
+		{
+			if (m_stream->eof())
+			{
+				m_file_done = true;
+				return;
+			}
+			std::getline(*m_stream, m_current_line);
+			m_line_number++;
+			if (!arff_detail::is_blank(m_current_line))
+				func();
+		}
+#ifndef SWIG
+		/**
+		 * Checks if a token represented by a string
+		 * denotes a primitive type in the ARFF format
+		 * @param token the token to be checked
+		 * @return whether the token denotes a primitive type
+		 */
+		SG_FORCED_INLINE bool is_primitive_type(const std::string& token) const
+		    noexcept
+		{
+			return token.find_first_of("numeric") != std::string::npos ||
+			       token.find_first_of("integer") != std::string::npos ||
+			       token.find_first_of("real") != std::string::npos ||
+			       token.find_first_of("string") != std::string::npos;
+		}
+#endif // SWIG
+
+		void reserve_vector_memory(size_t line_count)
+		{
+			VectorResizeVisitor visitor{line_count};
+			for (auto& vec : m_data_vectors)
+				shogun::visit(visitor, vec);
+		}
+
+		/**
+		 * Visitor pattern to reserve memory for a std::vector
+		 * wrapped in a variant class.
+		 */
+		struct VectorResizeVisitor
```

Comment:
I guess this should be hidden from SWIG as well?

Reply:
ah sorry this is private...why does the above need to be hidden then?

Reply:
it doesn't, I should even move this to the cpp file